provider/azure/environ.go

```diff
@@ -1765,41 +1786,142 @@ func (env *azureEnviron) deleteResourceGroup(ctx context.ProviderCallContext, sd
+
+	// Find all the resources tagged as belonging to this model.
+	resourcesClient := resources.Client{env.resources}
+	filter := fmt.Sprintf("tagName eq '%s' and tagValue eq '%s'", tags.JujuModel, env.config.UUID())
+	result, err := resourcesClient.ListByResourceGroupComplete(sdkCtx, resourceGroup, filter, "", nil)
 	if err != nil {
-		return errors.Trace(err)
+		return errors.Annotate(err, "listing resources to delete")
 	}
-	deployment := resources.Deployment{
-		Properties: &resources.DeploymentProperties{
-			Mode:     resources.Complete,
-			Template: &templateMap,
-		},
+
+	var resourceItems []resources.GenericResourceExpanded
+	for result.NotDone() {
+		resourceItems = append(resourceItems, result.Value())
+		err = result.NextWithContext(sdkCtx)
+		if err != nil {
+			return errors.Trace(err)
+		}
 	}
-	future, err := client.CreateOrUpdate(
-		sdkCtx,
-		resourceGroup,
-		"purge-resource-group",
-		deployment,
+
+	// Older APIs can ignore the filter above, so query the hard way just in case.
+	if len(resourceItems) == 0 {
+		result, err = resourcesClient.ListByResourceGroupComplete(sdkCtx, resourceGroup, "", "", nil)
+		if err != nil {
+			return errors.Annotate(err, "listing resources to delete")
+		}
+		for result.NotDone() {
+			res := result.Value()
+			fullRes, err := resourcesClient.GetByID(sdkCtx, to.String(res.ID), computeAPIVersion)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			if env.config.UUID() != to.String(fullRes.Tags[tags.JujuModel]) {
+				continue
+			}
+			resourceItems = append(resourceItems, res)
+			err = result.NextWithContext(sdkCtx)
+			if err != nil {
+				return errors.Trace(err)
+			}
+		}
+	}
+
+	// These will be deleted as part of stopping the instance below.
+	machineResourceTypes := set.NewStrings(
+		"Microsoft.Compute/virtualMachines",
+		"Microsoft.Compute/disks",
+		"Microsoft.Network/publicIPAddresses",
+		"Microsoft.Network/networkInterfaces",
 	)
-	if err != nil {
-		return errorutils.HandleCredentialError(errors.Annotatef(err, "purging resource group %q", resourceGroup), ctx)
+
+	var ids []instance.Id
+	var otherResources []resources.GenericResourceExpanded
+	for _, r := range resourceItems {
+		rType := to.String(r.Type)
+		logger.Debugf("resource to delete: %v (%v)", to.String(r.Name), rType)
+		if !machineResourceTypes.Contains(rType) {
+			otherResources = append(otherResources, r)
+			continue
+		}
+		if rType == "Microsoft.Compute/virtualMachines" {
+			ids = append(ids, instance.Id(to.String(r.Name)))
+		}
 	}
 
-	err = future.WaitForCompletionRef(sdkCtx, client.Client)
+	// Stopping instances will also remove most of their dependent resources.
+	err = env.StopInstances(ctx, ids...)
 	if err != nil {
-		return errors.Annotatef(err, "purging resource group %q", resourceGroup)
+		return errors.Annotatef(err, "deleting machine instances %q", ids)
+	}
+
+	// Loop until all remaining resources are deleted.
+	remainingResources := otherResources
+	for len(remainingResources) > 0 {
```

Comment:
It it worth putting a ceiling on how long we will keep trying this?

Reply:
Add for safety, in reality it will be at most 2 passes